### PR TITLE
Loads secrets in parallel

### PR DESCRIPTION
**Pull request by pakrym**
Fixes: https://github.com/aspnet/Extensions/issues/910

### CODE REVIEW

File: src/Configuration/Config.AzureKeyVault/test/AzureKeyVaultConfigurationTest.cs [hunk 1]
```diff
@@ -170,6 +171,50 @@ public void ReplaceDoubleMinusInKeyName()
...
+
+            client.Setup(c => c.GetSecretAsync(It.IsAny<string>()))
+                .Returns(async (string id) => {
+                    var shortId = id.Substring(id.LastIndexOf('/') + 1);
+                    if (Interlocked.Decrement(ref expectedCount) == 0)
+                    {
+                        tcs.SetResult(null);
+                    }
+
+                    await tcs.Task;
```
Comment (BrennanConroy) [inline src/Configuration/Config.AzureKeyVault/test/AzureKeyVaultConfigurationTest.cs]: nit: use the `WithTimeout` helper functions here

**Comment by pakrym** [comment]
This is technically a breaking change. We now require thread safety from customer provided `IKeyVaultSecretManager` implementations where it wasn't required before.

On the other hand it's inlikely that `IKeyVaultSecretManager` implementations would contain any state non-thread safe state considering it's purpose is filtering and transforming secret names into configuration key names.

``` C#
/// <summary>
    /// The <see cref="IKeyVaultSecretManager"/> instance used to control secret loading.
    /// </summary>
    public interface IKeyVaultSecretManager
    {
        /// <summary>
        /// Checks if <see cref="SecretItem"/> value should be retrieved.
        /// </summary>
        /// <param name="secret">The <see cref="SecretItem"/> instance.</param>
        /// <returns><code>true</code> is secrets value should be loaded, otherwise <code>false</code>.</returns>
        bool Load(SecretItem secret);

        /// <summary>
        /// Maps secret to a configuration key.
        /// </summary>
        /// <param name="secret">The <see cref="SecretBundle"/> instance.</param>
        /// <returns>Configuration key name to store secret value.</returns>
        string GetKey(SecretBundle secret);
    }

```

/cc @muratg @Eilon

**Comment by Eilon** [comment]
Out of curiosity, with this change, what is the time delta in the scenario of ~79 secrets being loaded?

**Comment by Eilon** [comment]
As far as breaking changes go, I agree with the assessment. Can we update the doc comments for the affected types/methods to indicate the thread safety requirements?

File: src/Configuration/Config.AzureKeyVault/src/AzureKeyVaultConfigurationProvider.cs [hunk 2]
```diff
@@ -52,16 +54,21 @@ private async Task LoadAsync()
             var secrets = await _client.GetSecretsAsync(_vault).ConfigureAwait(false);
             do
             {
+                var tasks = new List<Task<KeyValuePair<string, string>>>();
```
Comment (gfoidl) [inline src/Configuration/Config.AzureKeyVault/src/AzureKeyVaultConfigurationProvider.cs]: Can the List be initialized with the count of `secrets`?
Though in this case it's really minor...

File: src/Configuration/Config.AzureKeyVault/src/AzureKeyVaultConfigurationProvider.cs [hunk 3]
```diff
@@ -52,16 +54,21 @@ private async Task LoadAsync()
...
                 foreach (var secretItem in secrets)
                 {
-                    if (!_manager.Load(secretItem) || (secretItem.Attributes?.Enabled != true))
+                    if (!_manager.Load(secretItem) || secretItem.Attributes?.Enabled != true)
```
Comment (gfoidl) [inline src/Configuration/Config.AzureKeyVault/src/AzureKeyVaultConfigurationProvider.cs]: I think the "positive case" is more readable:

```c#
foreach (var secretItem in secrets)
{
    if (_manager.Load(secretItem) && secretItem.Attributes?.Enabled ?? false)
    {
        tasks.Add(LoadSecretValue(secretIItem);
    }
}
```

File: src/Configuration/Config.AzureKeyVault/src/AzureKeyVaultConfigurationProvider.cs [hunk 4]
```diff
@@ -52,16 +54,21 @@ private async Task LoadAsync()
...
                     {
                         continue;
                     }
+                    tasks.Add(LoadSecretValue(secretItem));
+                }
+
+                await Task.WhenAll(tasks);
```
Comment (gfoidl) [inline src/Configuration/Config.AzureKeyVault/src/AzureKeyVaultConfigurationProvider.cs]: `ConfigureAwait`?

File: src/Configuration/Config.AzureKeyVault/src/AzureKeyVaultConfigurationProvider.cs [hunk 5]
```diff
@@ -52,16 +54,21 @@ private async Task LoadAsync()
...
 
-                    var value = await _client.GetSecretAsync(secretItem.Id).ConfigureAwait(false);
-                    var key = _manager.GetKey(value);
-                    data.Add(key, value.Value);
+                foreach (var task in tasks)
+                {
+                    data.Add(task.Result.Key, task.Result.Value);
```
Comment (gfoidl) [inline src/Configuration/Config.AzureKeyVault/src/AzureKeyVaultConfigurationProvider.cs]:
```suggestion
                    var (key, value) = task.Result;
                    data.Add(key, value);
```
?

Offtopic: it's a bit strange to "deconsruct" a `KeyValuePair<K, V>` if you want to add it to a dictionary :wink:

File: src/Configuration/Config.AzureKeyVault/src/AzureKeyVaultConfigurationProvider.cs [hunk 6]
```diff
@@ -71,5 +78,12 @@ private async Task LoadAsync()
 
             Data = data;
         }
+
+        private async Task<KeyValuePair<string, string>> LoadSecretValue(SecretItem secretItem)
```
Comment (gfoidl) [inline src/Configuration/Config.AzureKeyVault/src/AzureKeyVaultConfigurationProvider.cs]: `LoadSecretValueAsync` as name?

**Comment by pakrym** [comment]
Not a breaking change anymore. All `IKeyVaultSecretManager` interactions are done on the same thread now.

**Comment by pakrym** [comment]
@Eilon our connection to KeyVault is pretty good so on my machine with 100 secrets load time went from 11 to 7 sec. I would imagine much better results on machines with slower access to keyvault.

**Pull request merged**